File: codility/lessons/3 - time complexity/tape equilibrium/tape_equilibrium.py
# Source: https://app.codility.com/programmers/lessons/3-time_complexity/tape_equilibrium/

def solution(array):

    first_part = array[0]
    second_part = sum(array) - first_part
    minimum_difference_between_parts = abs(first_part - second_part)

    for element in array[1:-1]:
        first_part += element
        second_part -= element
        CURRENT_MINIMUM = abs(first_part - second_part)
        minimum_difference_between_parts = min(minimum_difference_between_parts, CURRENT_MINIMUM)

    return minimum_difference_between_parts

File: codility/lessons/3 - time complexity/tape equilibrium/tape_equilibrium.go
// Source: https://app.codility.com/programmers/lessons/3-time_complexity/tape_equilibrium/

package solution

import "math"

func Solution(array []int) int {
    var firstPart int = array[0]
    var secondPart int = sumArrayElements(array) - firstPart
    var minimumDifferenceBetweenParts int = intAbs(firstPart, secondPart)

    for _, element := range array[1 : len(array)-1] {
        firstPart += element
        secondPart -= element
        var currentMinimum int = intAbs(firstPart, secondPart)
        minimumDifferenceBetweenParts = intMin(minimumDifferenceBetweenParts, currentMinimum)
    }

    return minimumDifferenceBetweenParts
}

    func sumArrayElements(array []int) int {
        sum := 0
        for _, element := range array {
            sum += element
        }
        return sum
    }

    func intAbs(num1 int, num2 int) int {
        return int(math.Abs(float64(num1) - float64(num2)))
    }

    func intMin(num1 int, num2 int) int {
        return int(math.Min(float64(num1), float64(num2)))
    }
